library/ssl_tls13_generic.c

```diff
@@ -1454,6 +1454,51 @@ int mbedtls_ssl_tls13_write_early_data_ext(mbedtls_ssl_context *ssl,
 
     return 0;
 }
+
+#if defined(MBEDTLS_SSL_SRV_C)
+int mbedtls_ssl_tls13_check_early_data_len(mbedtls_ssl_context *ssl,
+                                           size_t early_data_len)
+{
+    /*
+     * This function should be called only while an handshake is in progress
+     * and thus a session under negotiation. Add a sanity check to detect a
+     * misuse.
+     */
+    if (ssl->session_negotiate == NULL) {
+        return MBEDTLS_ERR_SSL_INTERNAL_ERROR;
+    }
+
+    /* RFC 8446 section 4.6.1
+     *
+     * A server receiving more than max_early_data_size bytes of 0-RTT data
+     * SHOULD terminate the connection with an "unexpected_message" alert.
+     */
+    if (early_data_len >
+        (ssl->session_negotiate->max_early_data_size -
+         ssl->total_early_data_size)) {
+
+        MBEDTLS_SSL_DEBUG_MSG(
+            2, ("EarlyData: Too much early data received, %u + %" MBEDTLS_PRINTF_SIZET " > %u",
+                ssl->total_early_data_size, early_data_len,
+                ssl->session_negotiate->max_early_data_size));
+
+        MBEDTLS_SSL_PEND_FATAL_ALERT(
+            MBEDTLS_SSL_ALERT_MSG_UNEXPECTED_MESSAGE,
+            MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE);
+        return MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE;
+    }
+
+    /*
+     * The check just above implies that early_data_len is lower than
+     * UINT32_MAX thus its cast to an uint32_t below is safe. We need it
+     * to appease some compilers.
+     */
```

Comment:
Sorry, I realise there's been a lot of changes to this comment, but I'm a bit uneasy with "the check just above". I can easily imagine the code above being changed and other checks added in the future, with this comment ignored, so that this becomes stale and confusing. Would it be better to say something like "The check against (uint32_t) max_early_data_size above...", or maybe even say something in the comment around line 1475 "Since max_early_data_size is uint32_t, this test means we can be sure early_data_len will fit in a uint32_t further down" with "We've checked that early_data_len will fit in a uint32_t further up"?

Reply:
Yes no worries, I agree with you. I will improve the comment, thanks.